util/repeatable_thread.h

```diff
@@ -0,0 +1,142 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include <functional>
+#include <string>
+
+#include "port/port.h"
+#include "rocksdb/env.h"
+#include "util/mutexlock.h"
+
+namespace rocksdb {
+
+class RepeatableThread {
+ public:
+  RepeatableThread(std::function<void()> function,
+                   const std::string& thread_name, Env* env, uint64_t delay_us,
+                   uint64_t initial_delay_us = 0)
+      : function_(function),
+        thread_name_("rocksdb:" + thread_name),
+        env_(env),
+        delay_us_(delay_us),
+        initial_delay_us_(initial_delay_us),
+        cond_var_(&mutex_),
+        running_(true),
+#ifndef NDEBUG
+        waiting_(false),
+        run_count_(0),
+#endif
+        thread_([this] { thread(); }) {
+  }
+
+  void cancel() {
+    {
+      MutexLock l(&mutex_);
+      if (!running_) {
+        return;
+      }
+      running_ = false;
+      cond_var_.SignalAll();
+    }
+    thread_.join();
+  }
+
+  ~RepeatableThread() { cancel(); }
+
+#ifndef NDEBUG
+  // Wait until RepeatableThread starting waiting, call the optional callback,
+  // then wait for one run of RepeatableThread. Tests can use provide a
+  // custom env object to mock time, and use the callback here to bump current
+  // time and trigger RepeatableThread. See repeatable_thread_test for example.
+  //
+  // Note: only support one caller of this method.
+  void TEST_WaitForRun(std::function<void()> callback = nullptr) {
+    MutexLock l(&mutex_);
+    while (!waiting_) {
+      cond_var_.Wait();
+    }
+    uint64_t prev_count = run_count_;
```

Comment:
I guess this is minor. Is mutex_ required while calling `callback`? Maybe should release the lock before and re-acquire later.

Reply:
Here I would expect caller use the callback in the same way as in repeatable_thread_test: bumping the time of a MockTimeEnv to trigger next run. In that case whether to unlock doesn't matter much.

Reply:
`waiting_` will never be `true` unless `mutex_` is release.
Because `wait()` need `mutex_` to change `waiting_` to `true`.
And `wait()` changes `waiting_` to `false` before release `mutex_`.

Thus `RepeatableThreadTest::MockEnvTest` will always hang forever.

Reply:
line 83 will release mutex, and the `waiting_` check at line 59 will then pass. The MockEnvTest stops timer at this point so the loop of line 82-87 cannot break.

Reply:
@solicomo you miss the fact that cond_var.wait() implicitly releases mutex before go to sleep and re-acquire the mutex after come back from sleep.

Reply:
@yiwu-arbug Thanks for your explanation.

I run `repeatable_thread_test` on a Windows machine, all tests passed.
But on my Mac, `RepeatableThreadTest::MockEnvTest` always hang forever.

I investigated it today and found:

`TEST_WaitForRun()` always hang on line 60 while `wait()` gets trapped into an endless loop on line 82-87. Because `env_->NowMicros()` always returns 0.

I guess the reason is:

- `wait()` thread always gets the opportunity to run.
- we are using `MockEnv`, if `TEST_WaitForRun()` has no chance to call `mock_env_->set_current_time(i)`, time will never elapse, thus the loop on line 82-87 will be infinite.

To fix it, we can use default env instead of `MockEnv`.

Reply:
@solicomo Thank you for your feedback. Very appreciate it. I come up with the `MockEnvTest` and `TEST_WaitForRun()` in the hope someone use `RepeatableThread` can have a way to mock time functions and have unit test that's not flaky. The `TimedTest` is intended to be an example of the same test with default env. Let me try to find a way to avoid the `TEST_WaitForRun()` being starve.

Reply:
@solicomo #4560